### PR TITLE
feat(sessds): preserve acks / replays in session state

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-9745](https://emqx.atlassian.net/browse/EMQX-9745).

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at f30b04f</samp>

This pull request enhances the persistent message and session features of emqx, by introducing a new API and data structure for replaying inflight messages, refactoring the session data structure and functions to use topic filters and subscriptions, and improving the performance and correctness of offset and range management. It also cleans up some unused or obsolete code in `emqx_persistent_session_ds.hrl`.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] ~~Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~~
- [x] Schema changes are backward compatible

### CODE REVIEW

File: apps/emqx/test/emqx_persistent_session_SUITE.erl [hunk 1]
```diff
@@ -555,6 +566,101 @@ t_publish_while_client_is_gone_qos1(Config) ->
...
+    %     #{payload := <<"M6">>}
+    % ], Msgs1),
+
+    NAcked = 4,
+    [ok = emqtt:puback(Client1, PktId) || #{packet_id := PktId} <- lists:sublist(Msgs1, NAcked)],
+
+    ok = emqtt:disconnect(Client1),
+    maybe_kill_connection_process(ClientId, Config),
+
+    NMsgs2 = 5,
```
Comment (ieQu1) [inline apps/emqx/test/emqx_persistent_session_SUITE.erl:629]: Better refactor it like this:
```
MSGS = [#mqtt_msg{...}, ...],
publish(MSGS),
...

receive_messages(length(MSGS), ...)
```

File: apps/emqx/src/emqx_persistent_message_ds_replayer.erl [hunk 2]
```diff
@@ -165,75 +152,195 @@ poll(SessionId, Inflight0, WindowSize) when WindowSize > 0, WindowSize < 16#7fff
...
+        ?SESSION_PUBRANGE_TAB,
+        #ds_pubrange{id = {SessionId, '_'}, _ = '_'},
+        read
+    ),
+    lists:map(fun export_range/1, DSRanges).
+
+export_range(#ds_pubrange{
+    type = Type, id = {_, First}, until = Until, stream = StreamRef, iterator = It
+}) ->
+    #range{type = Type, stream = StreamRef, first = First, until = Until, iterator = It}.
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_message_ds_replayer.erl:178]: Q: Can we avoid these transformation?

File: apps/emqx/test/emqx_persistent_session_SUITE.erl [hunk 3]
```diff
@@ -555,6 +566,101 @@ t_publish_while_client_is_gone_qos1(Config) ->
 
     ok = emqtt:disconnect(Client2).
 
+t_publish_many_while_client_is_gone_qos1(Config) ->
```
Comment (ieQu1) [inline apps/emqx/test/emqx_persistent_session_SUITE.erl:569]: Please add a comment describing what the scenario does

File: apps/emqx/src/emqx_persistent_message_ds_replayer.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx/src/emqx_persistent_message_ds_replayer.erl:178]: 👍🏼 Should be doable.

File: apps/emqx/src/emqx_persistent_session_ds.hrl [hunk 4]
```diff
@@ -34,25 +34,45 @@
...
+-type ds_pubrange() :: #ds_pubrange{}.
 
 -record(session, {
     %% same as clientid
     id :: emqx_persistent_session_ds:id(),
     %% creation time
     created_at :: _Millisecond :: non_neg_integer(),
     expires_at = never :: _Millisecond :: non_neg_integer() | never,
-    inflight :: emqx_persistent_message_ds_replayer:inflight(),
+    % last_ack = 0 :: emqx_persistent_message_ds_replayer:seqno(),
```
[the remark this answers is not in the document]
Comment (keynslug) [reply apps/emqx/src/emqx_persistent_session_ds.hrl:75]: Good point. I'm still unsure about it, so I'll drop it from this PR for now.

**Review by ieQu1 — APPROVED** [review]

**Pull request merged**